stdlib/LinearAlgebra/src/diagonal.jl

```diff
@@ -172,7 +172,7 @@ end
 
 function rmul!(A::AbstractMatrix, D::Diagonal)
     require_one_based_indexing(A)
-    A .= A .* transpose(D.diag)
+    A .= A .* permutedims(D.diag)
```

Comment:
`permutedims` here will allocate a new array since it is not lazy like `transpose`, right? Can that be avoided?

Reply:
Ah, good point. The typical answer would be `PermutedDimsArray`, but it doesn't support permuting non-existent dimensions.  I think a `reshape(D.diag, 1, :)` is the easiest solution.

Reply:
AFAICT, `permutedims(::AbstractVector)` does exactly that (that's also what the docstring says):

https://github.com/JuliaLang/julia/blob/d60503a00edfa19d6b13f8aba8ff3e0e4cdf9ae0/base/permuteddimsarray.jl#L178

It's the more generic `AbstractArray` method that allocates. And `D.diag` is guaranteed to be an `AbstractVector` by the `Diagonal{T,V}` parameter.